### DISCUSSION TITLE
Model hardstops in custom block

### DISCUSSION THREAD

**Discussion by kohn12** [OpenModelica/OpenModelica#11861] [Q&A]
I want to customize the behavior of a spring mechanism and use this to determine the position of a prismatic joint. The behavior can be seen as a pre-loaded spring with hard stops. When a force is applied to the mechanism, and the force is lower than the pre-load force, the displacement is zero and the force is transfered rigidly. If a force higher than the pre-load force is applied, the spring moves until it hits a hard stop. When moving, the displacement is found with the spring constant. When the hardstop is hit, the force is transferred rigidly again. This can be summarised with the following if-statement:
```modelica
if F_in < F_preload
  s = 0;
  f = Fin;
elseif (F_in-F_preload)/k < LimUp then
  s = (F_in-F_preload)/k ;
  f = 0;
else
  s = LimUp;
  f = F_in - k*Lim - F_preload;  
end if;
```
A custom block has been made with this if-statement as the basis of it. I looked into the (instantiated) code of 'Spring', 'ElastoGap' and 'MassWithStopAndFriction'. However when I run my code it is said to be overdetermined. This is probably due to the positions of the flanges being defined multiple times. What can i do to solve this, or what what would be another way to add hardstops in OpenModelica?

The code for my custom block can be seen below:
```modelica
block ElastoGapTest
extends Modelica.Mechanics.Translational.Interfaces.PartialCompliant;  
import Modelica.Units.SI;

// Parameters
parameter SI.TranslationalSpringConstant c(min=0, start=100) "Spring constant";
parameter SI.TranslationalDampingConstant d(min=0, start=1) "Damping constant";  
parameter SI.Position s_rel0 = 0 "Unstretched spring length";
constant Real LimUp = 0.1;
constant Real FVoor = 1000;

// Variables
Real Fin;
Real FLim;
Real f;
Real s_tijd;
Real s;

 equation
// Input force
Fin = flange_a.f;
FLim = c * LimUp;
s_tijd = (abs(Fin) - FVoor) / c;

// Conditional logic
if abs(Fin) < FVoor then
s = 0;
f = Fin;
  elseif s_tijd < LimUp then
s = s_tijd;
f = 0;
else
s = LimUp;
f = abs(Fin) - FLim - FVoor;  
end if;

// Negative if negative input
if Fin < 0 then
s = -s;
f = -f;
end if;

end ElastoGapTest;
```
My custom block is used in the following code:
```modelica
model TestModel
  inner Modelica.Mechanics.MultiBody.World world 
  Modelica.Mechanics.MultiBody.Joints.Revolute revolute 
  Modelica.Mechanics.MultiBody.Parts.Body body(m = 10) 
  Modelica.Mechanics.MultiBody.Parts.BodyBox bodyBox(r = {1, 0, 0})  
  Modelica.Mechanics.MultiBody.Joints.Prismatic prismatic(useAxisFlange = true)  
  ElastoGapTest elastoGapTest
equation
  connect(world.frame_b, revolute.frame_a)
  connect(bodyBox.frame_a, revolute.frame_b)
  connect(prismatic.frame_b, body.frame_a)
  connect(prismatic.frame_a, bodyBox.frame_b)
  connect(prismatic.support, elastoGapTest.flange_a)
  connect(prismatic.axis, elastoGapTest.flange_b)
end TestModel;
```

**Comment by phannebohm** [comment]
Does your model work if you remove the if-equation
```modelica
// Negative if negative input
if Fin < 0 then
  s = -s;
  f = -f;
end if;
```
or make it an `elseif` combined with the if-equation above?

BTW, `s = -s` is a cumbersome way to say `s = 0.0`.

**Comment by casella** [comment]
@kohn12, you hardstop model cannot work in OpenModelica and in other mainstream Modelica tools. The reason is that as soon as you connect it to a body with inertia, it will generate a variable-index system, where some differentiated variables (s, in this case) are constrained to have some values in some cases, but free to behave as proper states in others.

Handling these variable-index systems is a hard problem. People have tried ways to solve it, see, e.g., Dirk Zimmer's PhD thesis or John Tinnerholm's work on OpenModelica.jl. But it's still an open question how to do this efficiently.

You should model hard stops so that they are very stiff, but not infinitely stiff. And of course provide them with some damping, otherwise they may bounce like crazy. Not a completely trivial issue either, but doable.
